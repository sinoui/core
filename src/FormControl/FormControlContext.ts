import React from 'react';

export interface FormControlContextData {
  /**
   * 指定表单控件的id
   */
  id?: string;
  /**
   * 输入框形态
   */
  variant?: 'standard' | 'filled' | 'outlined';
  /**
   * 指定标签的布局方式
   */
  labelLayout?: 'floating' | 'shrink' | 'standard';
  /**
   * 表单控件是否已填充值
   */
  filled?: boolean;
  /**
   * 如果设置为`true`，则表示此表单控件必填。
   */
  required?: boolean;
  /**
   * 表单控件处于校验错误状态。
   */
  error?: boolean;
  /**
   * 设置为`true`，表示不可用。
   */
  disabled?: boolean;
  /**
   * 设置为`true`，表示表单控件处于聚焦状态。
   */
  focused?: boolean;
  /**
   * 标签是否显示冒号
   */
  colon?: boolean;
  /**
   * 输入框聚焦回调函数
   */
  onFocus(): void;
  /**
   * 输入框失去焦点时的回调函数
   */
  onBlur(): void;
  /**
   * 密集模式
   */
  dense?: boolean;
  /**
   * 标签元素引用
   */
  labelRef?: React.RefObject<HTMLLabelElement>;
  /**
   * 表单控件的布局模式
   */
  layout?: 'horizontal' | 'vertical';
}

const FormControlContext = React.createContext<FormControlContextData | null>(
  null,
);

export default FormControlContext;
