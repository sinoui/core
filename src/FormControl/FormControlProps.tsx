export default interface FormControlProps
  extends React.ComponentPropsWithRef<'div'> {
  /**
   * 设置标签。
   */
  label?: React.ReactNode;
  /**
   * 设置表单控件帮助性文本。
   */
  helperText?: React.ReactNode;
  /**
   * 表单控件校验错误信息。
   */
  error?: React.ReactNode;
  /**
   * 指定标签与表单控件之间的布局模式。默认为垂直布局（`vertical`）。`horizontal`表示水平布局。
   */
  layout?: 'horizontal' | 'vertical';
  /**
   * 指定表单控件的id
   */
  id?: string;
  /**
   * 输入框形态
   */
  variant?: 'standard' | 'filled' | 'outlined';
  /**
   * 指定标签的布局方式
   */
  labelLayout?: 'floating' | 'shrink' | 'standard';
  /**
   * 表单控件是否已填充值
   */
  filled?: boolean;
  /**
   * 如果设置为`true`，则表示此表单控件必填。
   */
  required?: boolean;
  /**
   * 设置为`true`，表示不可用。
   */
  disabled?: boolean;
  /**
   * 设置为`true`，表示表单控件处于聚焦状态。
   */
  focused?: boolean;
  /**
   * 标签是否显示冒号
   */
  colon?: boolean;
  /**
   * 密集模式
   */
  dense?: boolean;
  /**
   * 标签宽度 只有水平布局支持此属性
   */
  labelWidth?: number;
}
