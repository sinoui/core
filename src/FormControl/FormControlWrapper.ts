import styled, { css } from 'styled-components';
import type FormControlProps from './FormControlProps';

const formControlHorizontalDenseStyle = css<FormControlProps>`
  &.sinoui-form-item--horizontal > .sinoui-form-label {
    min-height: 32px;
    padding-top: 4px;
  }
`;
const FormControlWrapper = styled.div<FormControlProps>`
  display: flex;
  position: relative;
  flex-direction: column;
  align-items: stretch;

  .sinoui-form-item__content {
    display: flex;
    flex-direction: column;
  }

  @media screen and (min-width: ${({ theme }) => theme.breakpoints.sm}px) {
    &.sinoui-form-item--horizontal {
      flex-direction: row;
      align-items: flex-start;
    }

    &.sinoui-form-item--horizontal > .sinoui-form-item__content {
      flex: 1 1 0%;
    }

    &.sinoui-form-item--horizontal > .sinoui-form-label {
      display: flex;
      justify-content: flex-end;
      align-self: flex-start;
      font-size: 16px;
      text-align: right;
      width: ${({ labelWidth = 160 }) => labelWidth}px;
      min-height: 40px;
      padding-top: 8px;
      padding-right: 8px;
      box-sizing: border-box;

      &::after {
        content: ':';
      }
    }
  }

  ${(props) => props.dense && formControlHorizontalDenseStyle}
`;

export default FormControlWrapper;
