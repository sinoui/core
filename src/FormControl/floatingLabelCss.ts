import { css } from 'styled-components';
import FormLabelProps from './FormLabelProps';

const getTranslateX = ({
  variant,
}: {
  variant?: 'standard' | 'filled' | 'outlined';
}) => {
  switch (variant) {
    case 'filled':
      return 12;
    case 'outlined':
      return 14;

    case 'standard':
    default:
      return 0;
  }
};

const getTranslateY = (props: FormLabelProps) => {
  const { variant, filled, focused, dense } = props;
  const isShrink = filled || focused;
  if (variant === 'filled') {
    if (dense) {
      return isShrink ? 7 : 17;
    }
    return isShrink ? 9 : 20;
  }

  if (variant === 'outlined') {
    if (isShrink) {
      return -6;
    }
    return dense ? 12 : 20;
  }

  return isShrink ? 1.5 : 24;
};

const getScale = (props: FormLabelProps) => {
  const { filled, focused } = props;
  const isShrink = filled || focused;
  return isShrink ? 0.75 : 1;
};

/**
 * 浮动标签的样式
 */
const floatingLabelCss = css`
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1;
  pointer-events: none;
  transform-origin: top left;
  line-height: 1.3;
  transform: translate(${getTranslateX}px, ${getTranslateY}px)
    scale(${getScale});
  will-change: transform;
  max-width: calc(100% - 24px);
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export default floatingLabelCss;
