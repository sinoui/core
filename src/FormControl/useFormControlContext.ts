import { useContext } from 'react';
import FormControlContext from './FormControlContext';

/**
 * 获取表单控件上下文
 */
function useFormControlContext() {
  const formControlContext = useContext(FormControlContext);

  return formControlContext;
}

export default useFormControlContext;
