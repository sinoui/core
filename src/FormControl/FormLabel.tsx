import styled, { css } from 'styled-components';
import type { Theme } from '@sinoui/theme';
import React from 'react';
import classNames from 'classnames';
import mem from '../utils/mem';
import useFormControlContext from './useFormControlContext';
import type FormLabelProps from './FormLabelProps';
import floatingLabelCss from './floatingLabelCss';
import { FormControlContextData } from './FormControlContext';

/**
 * 获取标签颜色
 */
const getColor = ({
  theme,
  color = theme.palette.primary.main,
  focused,
  disabled,
  error,
}: FormLabelProps & { theme: Theme }) => {
  if (error) {
    return theme.palette.error.main;
  }
  if (disabled) {
    return theme.palette.text.disabled;
  }
  if (focused) {
    return color;
  }
  return theme.palette.text.secondary;
};

const requiredCss = css<{ $isAfterRequired: boolean }>`
  &::${({ $isAfterRequired }) => ($isAfterRequired ? 'after' : 'before')} {
    content: '*';
    padding: 4px;
  }
`;

const shrinkCss = css`
  font-size: ${({ theme }) => theme.typography.caption.fontSize};
  padding-bottom: 4px;
`;

const standardCss = css`
  padding-bottom: 4px;
`;

type FormLabelInnerPropsType = FormLabelProps & { $isAfterRequired: boolean };

const FormLabelInner = styled.label.attrs(
  ({
    className,
    variant,
    layout,
    $isAfterRequired,
    filled,
    focused,
  }: FormLabelInnerPropsType) => ({
    className: classNames(
      className,
      'sinoui-form-label',
      `sinoui-form-label--variant-${variant ?? 'standard'}`,
      `sinoui-form-label--layout-${layout ?? 'standard'}`,
      {
        'sinoui-form-label--focused': focused,
        'sinoui-form-label--filled': filled,
        'sinoui-form-label--required-is-after': $isAfterRequired,
      },
    ),
  }),
)<FormLabelInnerPropsType>`
  font-size: ${(props) => props.theme.typography.body1.fontSize};
  font-family: ${(props) => props.theme.typography.fontFamily};
  font-weight: ${(props) => props.theme.typography.fontWeightRegular};
  color: ${getColor};
  box-sizing: border-box;
  margin: 0px;
  margin-top: -2px;
  transition: ${({ theme }) =>
    theme.transitions.create(['color', 'transform'], {
      easing: theme.transitions.easing.easeInOut,
    })};

  ${({ required }) => required && requiredCss};
  ${({ layout = 'shrink' }) => layout === 'shrink' && shrinkCss}
  ${({ layout }) => layout === 'standard' && standardCss}
  ${({ layout }) => layout === 'floating' && floatingLabelCss}
`;

const mapper: Array<
  | (keyof FormControlContextData & keyof FormLabelProps)
  | [keyof FormControlContextData, keyof FormLabelProps]
> = [
  'error',
  'colon',
  'variant',
  'filled',
  'required',
  'disabled',
  'focused',
  'dense',
  ['labelLayout', 'layout'],
  ['id', 'htmlFor'],
];

/**
 * 从上下文中获取状态
 *
 * @param formControlContext 表单控件上下文
 */
const getStateFromContext = mem(
  (formControlContext?: FormControlContextData | null) => {
    const state: Partial<FormLabelProps> = {};
    if (!formControlContext) {
      return state;
    }
    mapper.forEach((item) => {
      const fromKey = Array.isArray(item) ? item[0] : item;
      const toKey = Array.isArray(item) ? item[1] : item;
      const value: any = formControlContext[fromKey];
      if (value != null) {
        state[toKey] = value;
      }
    });

    return state;
  },
);

/**
 * 表单标签组件
 *
 * @ReactComponent
 */
const FormLabel = React.forwardRef<HTMLLabelElement, FormLabelProps>(
  ({ children, ...props }, ref) => {
    const formControlContext = useFormControlContext();

    const labelProps = {
      ...getStateFromContext(formControlContext),
      ...props,
      ref,
      $isAfterRequired: formControlContext?.layout !== 'horizontal',
    };

    return React.createElement(FormLabelInner, labelProps, children);
  },
);

export default FormLabel;
