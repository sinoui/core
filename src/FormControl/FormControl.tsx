import React, { useState, useCallback, useRef } from 'react';
import HelperText from '../HelperText';
import bemClassNames from '../utils/bemClassNames';
import useId from './useId';
import HelperLine from '../HelperLine';
import type FormControlProps from './FormControlProps';
import FormControlWrapper from './FormControlWrapper';
import FormLabel from '../FormLabel';
import FormControlContext from './FormControlContext';

/**
 * 表单控件组件
 */
const FormControl = React.forwardRef<HTMLDivElement, FormControlProps>(
  (props, ref) => {
    const inputId = useId('input');
    const labelRef = useRef<HTMLLabelElement>(null);
    const {
      children,
      label,
      helperText,
      error,
      layout = 'vertical',
      className,
      id = inputId,
      filled,
      variant = 'standard',
      labelLayout,
      colon = true,
      disabled,
      required,
      focused: focusedProp,
      dense,
      ...rest
    } = props;

    const [_focused, setFocused] = useState(false);
    const focused = focusedProp ?? _focused;
    const onFocus = useCallback(() => setFocused(true), [setFocused]);
    const onBlur = useCallback(() => setFocused(false), [setFocused]);

    const childContext = {
      filled,
      id,
      variant,
      labelLayout,
      error: !!error,
      colon,
      disabled,
      required,
      focused,
      onFocus,
      onBlur,
      labelRef,
      dense,
      layout,
    };

    return (
      <FormControlContext.Provider value={childContext}>
        <FormControlWrapper
          className={bemClassNames('sinoui-form-item', {
            horizontal: layout === 'horizontal',
            vertical: layout === 'vertical',
            floating: labelLayout === 'floating',
          })}
          ref={ref}
          dense={dense}
          colon={colon}
          {...rest}
        >
          <FormLabel ref={labelRef}>{props.label}</FormLabel>
          <div className="sinoui-form-item__content">
            {props.children}
            <HelperLine>
              {helperText && !error && <HelperText>{helperText}</HelperText>}
              {error && <HelperText error>{error}</HelperText>}
            </HelperLine>
          </div>
        </FormControlWrapper>
      </FormControlContext.Provider>
    );
  },
);

export default FormControl;
