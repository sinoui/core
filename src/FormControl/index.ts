import FormControl from './FormControl';
import type FormControlProps from './FormControlProps';

export { default as useFormControlContext } from './useFormControlContext';
export type { FormControlProps };
export default FormControl;
