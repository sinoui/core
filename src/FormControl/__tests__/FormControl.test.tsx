import { render, cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import React from 'react';
import '@testing-library/jest-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import FormControl from '../FormControl';
import { useFormControlContext } from '..';
import type { FormControlContextData } from '../FormControlContext';

jest.mock('../useId', () => () => 'input_1');

afterAll(() => jest.unmock('../useId'));

afterEach(cleanup);

it('渲染FormControl', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormControl label="姓名" data-testid="formcontrol" helperText="必填">
        <input />
      </FormControl>
    </ThemeProvider>,
  );

  const formcontrol = getByTestId('formcontrol');

  expect(formcontrol.querySelector('label')).toHaveTextContent('姓名');
  expect(formcontrol.querySelector('.sinoui-helper-text')).toHaveTextContent(
    '必填',
  );
});

it('渲染校验错误信息', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormControl label="姓名" data-testid="formcontrol" error="必填">
        <input />
      </FormControl>
    </ThemeProvider>,
  );

  const formcontrol = getByTestId('formcontrol');

  expect(formcontrol.querySelector('label')).toHaveTextContent('姓名');
  expect(formcontrol.querySelector('.sinoui-helper-text')).toHaveTextContent(
    '必填',
  );
});

it('水平布局FormLabel显示冒号', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <FormControl
          label="姓名"
          data-testid="formcontrol"
          error="必填"
          layout="horizontal"
          id="input_1"
        >
          <input />
        </FormControl>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('ref指向根元素', () => {
  const ref = React.createRef<HTMLDivElement>();
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormControl data-testid="formcontrol" ref={ref}>
        <input />
      </FormControl>
    </ThemeProvider>,
  );

  expect(ref.current).toBe(getByTestId('formcontrol'));
});

it('className', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormControl data-testid="formcontrol">
        <input />
      </FormControl>
    </ThemeProvider>,
  );

  expect(getByTestId('formcontrol')).toHaveClass(
    'sinoui-form-item',
    'sinoui-form-item--vertical',
  );
});

it('自定义className', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormControl data-testid="formcontrol" className="custom-className">
        <input />
      </FormControl>
    </ThemeProvider>,
  );

  expect(getByTestId('formcontrol')).toHaveClass(
    'sinoui-form-item',
    'sinoui-form-item--vertical',
    'custom-className',
  );
});

it('labelLayout为floating时的class名称', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormControl data-testid="formcontrol" labelLayout="floating">
        <input />
      </FormControl>
    </ThemeProvider>,
  );

  expect(getByTestId('formcontrol')).toHaveClass(
    'sinoui-form-item',
    'sinoui-form-item--floating',
  );
});

it('水平布局', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormControl data-testid="formcontrol" layout="horizontal" id="input_1">
        <input />
      </FormControl>
    </ThemeProvider>,
  );

  expect(getByTestId('formcontrol')).toHaveClass(
    'sinoui-form-item--horizontal',
  );

  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <FormControl data-testid="formcontrol" layout="horizontal" id="input_1">
        <input />
      </FormControl>
    </ThemeProvider>,
  );

  expect(tree).toMatchSnapshot();
});

it('生成上下文', () => {
  let context: FormControlContextData | null | undefined;
  const Input = () => {
    context = useFormControlContext();
    return null;
  };

  render(
    <ThemeProvider theme={defaultTheme}>
      <FormControl colon required={false} error="校验错误" filled dense={false}>
        <Input />
      </FormControl>
    </ThemeProvider>,
  );

  expect(context?.colon).toBe(true);
  expect(context?.required).toBe(false);
  expect(context?.error).toBe(true);
  expect(context?.disabled).toBeUndefined();
  expect(context?.filled).toBe(true);
  expect(context?.labelLayout).toBeUndefined();
  expect(context?.variant).toBe('standard');
  expect(context?.focused).toBe(false);
  expect(context?.layout).toBe('vertical');
  expect(context?.labelRef).not.toBeUndefined();
  expect(context?.dense).toBe(false);

  act(() => {
    const onFocus = context?.onFocus;

    if (onFocus) {
      onFocus();
    }
  });

  expect(context?.focused).toBe(true);

  act(() => {
    const onBlur = context?.onBlur;

    if (onBlur) {
      onBlur();
    }
  });

  expect(context?.focused).toBe(false);
});
