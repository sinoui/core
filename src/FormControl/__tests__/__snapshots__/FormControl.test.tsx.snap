// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`水平布局 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  min-height: 24px;
  padding: 2px 12px;
  box-sizing: border-box;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  font-size: 0.75rem;
  line-height: 1.66;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c0 .sinoui-form-item__content {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .sinoui-helper-line {
  margin-bottom: 8px;
}

.c0 .sinoui-helper-line:empty {
  margin-bottom: 0;
}

.c1 {
  font-size: 1rem;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  font-weight: 400;
  color: rgba(0,0,0,0.6);
  box-sizing: border-box;
  margin: 0px;
  margin-top: -2px;
  -webkit-transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,-webkit-transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  font-size: 0.75rem;
  padding-bottom: 4px;
}

@media screen and (min-width:600px) {
  .c0.sinoui-form-item--horizontal {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-item__content {
    -webkit-flex: 1 1 0%;
    -ms-flex: 1 1 0%;
    flex: 1 1 0%;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-label {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
    -webkit-align-self: flex-start;
    -ms-flex-item-align: start;
    align-self: flex-start;
    font-size: 16px;
    text-align: right;
    width: 160px;
    min-height: 40px;
    padding-top: 8px;
    padding-right: 8px;
    box-sizing: border-box;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-label::after {
    content: ':';
  }
}

<div
  className="c0 sinoui-form-item sinoui-form-item--horizontal"
  data-testid="formcontrol"
>
  <label
    className="c1 sinoui-form-label sinoui-form-label--variant-standard sinoui-form-label--layout-standard"
    htmlFor="input_1"
  />
  <div
    className="sinoui-form-item__content"
  >
    <input />
    <div
      className="c2 sinoui-helper-line"
    />
  </div>
</div>
`;

exports[`水平布局FormLabel显示冒号 1`] = `
.c3 {
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  -webkit-letter-spacing: 0.03333em;
  -moz-letter-spacing: 0.03333em;
  -ms-letter-spacing: 0.03333em;
  letter-spacing: 0.03333em;
  color: #f44336;
  text-align: left;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  min-height: 24px;
  padding: 2px 12px;
  box-sizing: border-box;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  font-size: 0.75rem;
  line-height: 1.66;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c0 .sinoui-form-item__content {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .sinoui-helper-line {
  margin-bottom: 8px;
}

.c0 .sinoui-helper-line:empty {
  margin-bottom: 0;
}

.c1 {
  font-size: 1rem;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  font-weight: 400;
  color: #f44336;
  box-sizing: border-box;
  margin: 0px;
  margin-top: -2px;
  -webkit-transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,-webkit-transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  font-size: 0.75rem;
  padding-bottom: 4px;
}

@media screen and (min-width:600px) {
  .c0.sinoui-form-item--horizontal {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-item__content {
    -webkit-flex: 1 1 0%;
    -ms-flex: 1 1 0%;
    flex: 1 1 0%;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-label {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
    -webkit-align-self: flex-start;
    -ms-flex-item-align: start;
    align-self: flex-start;
    font-size: 16px;
    text-align: right;
    width: 160px;
    min-height: 40px;
    padding-top: 8px;
    padding-right: 8px;
    box-sizing: border-box;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-label::after {
    content: ':';
  }
}

<div
  className="c0 sinoui-form-item sinoui-form-item--horizontal"
  data-testid="formcontrol"
>
  <label
    className="c1 sinoui-form-label sinoui-form-label--variant-standard sinoui-form-label--layout-standard"
    htmlFor="input_1"
  >
    姓名
  </label>
  <div
    className="sinoui-form-item__content"
  >
    <input />
    <div
      className="c2 sinoui-helper-line"
    >
      <div
        className="c3 sinoui-helper-text"
      >
        必填
      </div>
    </div>
  </div>
</div>
`;
